rllib/algorithms/dqn/tf/dqn_tf_rl_module.py

```diff
@@ -0,0 +1,95 @@
+from typing import Any, Mapping
+
+from ray.rllib.algorithms.dqn.dqn_rl_module import DQNRLModule
+from ray.rllib.core.models.base import ENCODER_OUT
+from ray.rllib.core.rl_module.tf.tf_rl_module import TfRLModule
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.nested_dict import NestedDict
+
+_, tf, _ = try_import_tf()
+
+
+class DQNTfRLModule(TfRLModule, DQNRLModule):
+    @override
+    def setup(self) -> None:
+        super().setup(self)
+
+        # Primary update of the target network's weigths.
+        self.update_target_network()
+
+    @override(TfRLModule)
+    def _forward_inference(self, batch: NestedDict) -> Mapping[str, Any]:
+        output = {}
+
+        # State encoding.
+        encoder_outs = self.q_encoder(batch)
+
+        # TODO (simon): include RNN fucntionality.
+        # if STATE_OUT in encoder_outs:
+
+        # Q head.
+        q_outs = self.q(encoder_outs[ENCODER_OUT])
+        output[SampleBatch.ACTIONS] = tf.argmax(q_outs, axis=1)
+
+        return output
+
+    @override(TfRLModule)
+    def _forward_exploration(self, batch: NestedDict, **kwargs) -> Mapping[str, Any]:
+        output = {}
+
+        # State encoding.
+        encoder_outs = self.q_encoder(batch)
+
+        # TODO (simon): include RNN functionality.
+        # if STATE_OUT in encoder_outs:
+
+        # Q head.
+        q_outs = self.q(encoder_outs[ENCODER_OUT])
+
+        # Include epsilon-greedy exploration.
+        exploitation_actions = tf.argmax(q_outs, axis=1)
+        epsilon = kwargs.get("epsilon", 0.3)
+
+        # Draw random actions from categorical distributions
+        # based on the q logits.
+        random_actions = tf.squeeze(tf.random.categorical(q_outs, 1))
+        B = tf.shape(q_outs)
+        randomize = (
+            tf.random.uniform(tf.stack([B]), minval=0.0, maxval=1.0, dtype=tf.float32)
+            < epsilon
+        )
+        actions = tf.where(randomize, random_actions, exploitation_actions)
+        output[SampleBatch.ACTIONS] = actions
+
+        return output
+
+    @override(TfRLModule)
+    def _forward_train(self, batch: NestedDict) -> Mapping[str, Any]:
+        output = {}
+
+        # State encoding.
+        encoder_outs = self.q_encoder(batch)
+
+        # Q head.
+        q_outs = self.q(encoder_outs[ENCODER_OUT])
+
+        # TODO (simon): Either override the `Encoder` for the target
+        # or override the `input_specs_train` for the `target_encoder`
+        # to accept the `SampleBatch.NEXT_OBS` instead of the
+        # `SampleBatch.OBS`.
+
+        output["q_values"] = q_outs
+        return output
+
+    # TODO (kourosh, avnish): I see that there is an
```

Comment:
@kouroshHakha @avnishn Could you give me a feedback how to implement the target network update? 

I would call this function in the `Learner.compute_loss_for_module()`. In `APPO` the `RLModuleWithTargetNetworksInterface`  is providing pairs whioch get explicitly updated in the learner. 

The approach here would leave the specifics to the module to make its target network updates.

Reply:
hey @simonsays1980,

Let's adopt the same pattern as APPO. Basically target module in DQN is updated using polyak averaging. We have `Learner.additional_update_for_module()` API exactly for this purpose. Inside you can define an `_update_module_target_networks` api in which you can use `module.get_target_network_pairs()` to get the pairs of encoder and heads for main and target functions. 

Let me know if it's not clear.

Reply:
Using now the `RLModuleWithTargetNetworksInterface`. Will update it similar to `APPO` during the `Learner` calls to `additional_update_for_module()` using pairs and assignment.

Let me know if this fits your idea.

Reply:
@simonsays1980 Kourosh is OOO for a bit now. Please go ahead with the implementation as you described it.
Implement `get_target_network_pairs` to return the model pairs (RLModule) and `additional_update_for_module` (Learner) to define what should happen to these model pairs.

Reply:
@ArturNiederfahrenhorst Alright. Have done it in [this commit](https://github.com/simonsays1980/ray/commit/8a1d6ccfe06f440d3dd9506fe59fd410e25fa5d7)

For the next steps we have to check what we do with the policy.